source/common/filter/tcp_proxy.cc

```diff
@@ -56,14 +121,23 @@ void TcpProxy::initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callb
 }
 
 Network::FilterStatus TcpProxy::initializeUpstreamConnection() {
-  Upstream::ClusterInfoPtr cluster = cluster_manager_.get(config_->clusterName());
+  const std::string& cluster_name = config_->getClusterForConnection(read_callbacks_->connection());
+
+  Upstream::ClusterInfoPtr cluster = cluster_manager_.get(cluster_name);
+
+  if (cluster) {
+    conn_log_debug("Creating connection to cluster {}", read_callbacks_->connection(),
+                   cluster_name);
+  } else {
+    return Network::FilterStatus::StopIteration;
```

Comment:
Currently this can't happen, because you verify that all the clusters exist. If we eventually want to support CDS/RDS like constructs for tcp_proxy, we will need to deal with this, but for now I would not have this logic at all and just crash if cluster is nullptr.